crates/bitwarden-threading/src/thread_bound_dispatcher.rs

```diff
@@ -0,0 +1,182 @@
+#![allow(dead_code)]
+#![allow(unused_variables)]
+
+use std::{future::Future, pin::Pin, sync::Arc};
+
+use thiserror::Error;
+#[cfg(not(target_arch = "wasm32"))]
+use tokio::task::spawn_local;
+
+#[cfg(target_arch = "wasm32")]
+use wasm_bindgen_futures::spawn_local;
+
+struct CallRequest<ThreadState> {
+    function: Box<dyn FnOnce(Arc<ThreadState>) -> Pin<Box<dyn Future<Output = ()>>> + Send>,
+}
+
+#[derive(Debug, Error)]
+pub enum CallError {
+    #[error("Failed to request function call: {0}")]
+    ChannelSend(String),
+
+    #[error("Failed to receive return value: {0}")]
+    ChannelReceive(String),
+}
+
+pub struct ThreadBoundDispatcher<ThreadState> {
+    call_channel_tx: tokio::sync::mpsc::Sender<CallRequest<ThreadState>>,
+}
+
+impl<ThreadState> ThreadBoundDispatcher<ThreadState>
+where
+    ThreadState: 'static,
+{
+    pub fn new(state: ThreadState) -> Self {
+        let (call_channel_tx, mut call_channel_rx) =
+            tokio::sync::mpsc::channel::<CallRequest<ThreadState>>(1);
+
+        spawn_local(async move {
+            let state = Arc::new(state);
```

Comment:
The `Arc` never leaves this locally spawned task so we don't need the thread-safety and can safely use an `Rc` and avoid the overhead of the atomic operations.

Reply:
> I feel like there might even be a way to just pass &state

Haha oh yeah... that's what I started out with but the issue is (was?) that the reference crosses over multiple closure bounds which causes it to escape and make `rustc` sad. Your suggestion above might make it easier though. I'll try to give it a shot but it's probably not worth a massive increase in complexity

Reply:
Yeah figured that was the case for the references, we should still be able to swap `Arc` for `Rc` though.

Reply:
Done! 👍